File: pages/product_page.py
import re

from playwright.sync_api import Page, expect

from fixtures.test_data import SELECTORS
from pages.base_page import BasePage


class ProductPage(BasePage):
    def __init__(self, page: Page):
        super().__init__(page)
        self.product_title = page.locator(SELECTORS["product_title"]).first
        self.add_to_cart_button = page.locator(SELECTORS["add_to_cart"]).first
        self.price_label = page.locator('[class*="price"], .pdp-price').first

    def expect_loaded(self) -> None:
        self.page.wait_for_url(re.compile(r"products|-i\d+", re.I), timeout=45_000)
        self.dismiss_overlays()
        expect(self.product_title).to_be_visible(timeout=30_000)

    def expect_title_not_empty(self) -> None:
        title = self.product_title.text_content()
        assert title and title.strip()

    def expect_price_visible(self) -> None:
        expect(self.price_label).to_be_visible(timeout=20_000)

    def add_to_cart(self) -> None:
        self.dismiss_overlays()
        self.add_to_cart_button.scroll_into_view_if_needed()
        self.add_to_cart_button.click(timeout=20_000)

    def expect_add_to_cart_available(self) -> None:
        expect(self.add_to_cart_button).to_be_visible(timeout=25_000)

File: pages/product.page.ts
import { expect } from '@playwright/test';
import { BasePage } from './base.page';
import { SELECTORS } from '../fixtures/test-data';

export class ProductPage extends BasePage {
  private readonly productTitle = this.page.locator(SELECTORS.productTitle).first();
  private readonly addToCartButton = this.page.locator(SELECTORS.addToCart).first();
  private readonly priceLabel = this.page.locator('[class*="price"], .pdp-price').first();

  async expectLoaded() {
    await this.page.waitForURL(/products|-i\d+/i, { timeout: 45_000 });
    await this.dismissOverlays();
    await expect(this.productTitle).toBeVisible({ timeout: 30_000 });
  }

  async expectTitleNotEmpty() {
    const title = await this.productTitle.textContent();
    expect(title?.trim().length).toBeGreaterThan(0);
  }

  async expectPriceVisible() {
    await expect(this.priceLabel).toBeVisible({ timeout: 20_000 });
  }

  async addToCart() {
    await this.dismissOverlays();
    await this.addToCartButton.scrollIntoViewIfNeeded();
    await this.addToCartButton.click({ timeout: 20_000 });
  }

  async expectAddToCartAvailable() {
    await expect(this.addToCartButton).toBeVisible({ timeout: 25_000 });
  }
}
